services/rfq/relayer/pricer/fee_pricer.go

```diff
@@ -147,11 +171,92 @@ func (f *feePricer) GetDestinationFee(parentCtx context.Context, _, destination
 		fee = new(big.Int).Add(fee, l1Fee)
 		span.SetAttributes(attribute.String("l1_fee", l1Fee.String()))
 	}
+
 	span.SetAttributes(attribute.String("destination_fee", fee.String()))
 	return fee, nil
 }
 
-func (f *feePricer) GetTotalFee(parentCtx context.Context, origin, destination uint32, denomToken string, isQuote bool) (_ *big.Int, err error) {
+// addZapFees incorporates the cost of the call and the call value into the fee.
+// Note that to be conservative, we always use the QuoteFixedFeeMultiplier over the RelayFixedFeeMultiplier.
+//
+//nolint:gosec
+func (f *feePricer) addZapFees(ctx context.Context, destination uint32, denomToken string, quoteRequest *reldb.QuoteRequest, fee *big.Int) (*big.Int, error) {
+	span := trace.SpanFromContext(ctx)
+
+	if len(quoteRequest.Transaction.CallParams) > 0 {
+		gasEstimate, err := f.getZapGasEstimate(ctx, destination, quoteRequest)
+		if err != nil {
+			return nil, err
+		}
+		callFee, err := f.getFee(ctx, destination, destination, int(gasEstimate), denomToken, true)
+		if err != nil {
+			return nil, err
+		}
+		fee = new(big.Int).Add(fee, callFee)
+		span.SetAttributes(attribute.String("call_fee", callFee.String()))
+	}
+
+	if quoteRequest.Transaction.CallValue != nil && quoteRequest.Transaction.CallValue.Sign() > 0 {
+		callValueFloat := new(big.Float).SetInt(quoteRequest.Transaction.CallValue)
+		valueDenom, err := f.getDenomFee(ctx, destination, destination, denomToken, callValueFloat)
+		if err != nil {
+			return nil, err
+		}
+		valueScaled, err := f.getFeeWithMultiplier(ctx, destination, true, valueDenom)
+		if err != nil {
+			return nil, err
+		}
+		fee = new(big.Int).Add(fee, valueScaled)
+		span.SetAttributes(attribute.String("value_scaled", valueScaled.String()))
+	}
+
+	return fee, nil
+}
+
+// cache so that we don't have to parse the ABI every time.
+var fastBridgeV2ABI *abi.ABI
+
+const methodName = "relay0"
+
+func (f *feePricer) getZapGasEstimate(ctx context.Context, destination uint32, quoteRequest *reldb.QuoteRequest) (gasEstimate uint64, err error) {
+	client, err := f.clientFetcher.GetClient(ctx, big.NewInt(int64(destination)))
+	if err != nil {
+		return 0, fmt.Errorf("could not get client: %w", err)
+	}
+
+	if fastBridgeV2ABI == nil {
+		parsedABI, err := abi.JSON(strings.NewReader(fastbridgev2.IFastBridgeV2MetaData.ABI))
+		if err != nil {
+			return 0, fmt.Errorf("could not parse ABI: %w", err)
+		}
+		fastBridgeV2ABI = &parsedABI
+	}
+
+	encodedData, err := fastBridgeV2ABI.Pack(methodName, quoteRequest.RawRequest, f.relayerAddress)
+	if err != nil {
+		return 0, fmt.Errorf("could not encode function call: %w", err)
+	}
+
+	rfqAddr, err := f.config.GetRFQAddress(int(destination))
+	if err != nil {
+		return 0, fmt.Errorf("could not get RFQ address: %w", err)
+	}
+
+	callMsg := ethereum.CallMsg{
+		From:  f.relayerAddress,
+		To:    &rfqAddr,
+		Value: quoteRequest.Transaction.CallValue,
+		Data:  encodedData,
+	}
+	gasEstimate, err = client.EstimateGas(ctx, callMsg)
```

Comment:
What's the expected of `EstiamateGas` here if the transaction reverts to matter what? For instance another relayer might've relayed this sooner than us, causing the execution simulation to fail. I'm not sure if this throws an error or returns an absurdly big return value here.

Reply:
In my tests it returns an error. I think that's desired here? We could also have a fallback mechanism, though I'm not sure what that would look like.